### PR TITLE
[expo] Deprecate globals

### PR DESCRIPTION
**Pull request by brentvatne**
# Why

We want to remove these soon to clean up the expo package further.

# How

Added a warning similar to that of deprecatedModule

# Test Plan

Tested this out in an app by accessing a couple of properties on `global.__expo`:

Screen Shot 2020-12-08 at 5 44 42 PM

### CODE REVIEW

File: packages/expo/src/globals.ts [hunk 1]
```diff
@@ -6,47 +6,55 @@
 // the following RFC has been implemented:
 // https://github.com/react-native-community/discussions-and-proposals/issues/120
 
+import depreactedGlobal from './deprecatedGlobal';
```
Comment (ide) [inline packages/expo/src/globals.ts:9]:
```suggestion
import deprecatedGlobal from './deprecatedGlobal';
```

File: packages/expo/src/deprecatedGlobal.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+function logWarning() {
+  if (!packages.length) {
+    return;
+  }
+
+  let instructions = '';
+
+  packages = Array.from(new Set(packages));
+  packages.sort();
```
Comment (ide) [inline packages/expo/src/deprecatedGlobal.ts:24]: This isn't actually used

File: packages/expo/src/deprecatedGlobal.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  namedImports = Array.from(new Set(namedImports));
+  namedImports.sort();
+
+  extraInstructions = Array.from(new Set(extraInstructions));
+  extraInstructions.sort();
+
+  instructions += namedImports.join(', ');
+  instructions += `.\n\n`;
+  instructions += `If you are not using global.Expo or global.__expo directly in your source code, then they may used in one of your dependencies.\n`;
```
Comment (ide) [inline packages/expo/src/deprecatedGlobal.ts:34]:
```suggestion
  instructions += `If you are not using global.Expo or global.__expo directly in your source code, then they may used in one of your dependencies. The deprecated globals will be removed with expo v41.0.0.\n`;
```

**Review by ide — APPROVED** [review]

**Comment by Slapbox** [comment]
Might I suggest keeping a simple boolean like `window.Expo === true` or something?

Maybe we're an edge case, but the rationale for this being that we rely on `window.Expo` existing to determine whether or not we're running in React Native in our code. We have code that is shared with Electron, and when you open React Native code in React Native Debugger, [is-electron](https://www.npmjs.com/package/is-electron) is fooled into thinking that the code is running in Electron.

Is there some other React-Native global that already exists that we could rely on? Perhaps this is no longer an issue, but about 18 months ago when we set up this code, this was the only solution we could find that didn't involve doing the inverse and using `webpack.DefinePlugin` to set something like `window.isElectron` ourselves.

**Comment by brentvatne** [comment]
create `isExpo.js` and export `false` and then `isExpo.expo.js` and export `true`

**Comment by brentvatne** [comment]
oh actually another option could be to use the `.native.js` extension, which is loaded on ios and android in react native and expo apps but not in web/electron/whatever

**Comment by Slapbox** [comment]
@brentvatne thanks for your reply - how would this actually work on the non-React-Native side though? Webpack doesn't like trying to import files that don't exist. Do you know what the correct way to manage that is?

I figured I could either `exclude` with `exclude: [/\.native\.jsx?$/]` but no dice. 

So then I tried a negative lookbehind like `test: /(?<!\.native)\.jsx?$/`, but then I get unexpected errors that prevent me from running our project at all - in this case related in React Refresh,  but it seemed likely more were coming.

I then tried [the code advised on the deprecation page for .expo.js files](https://github.com/expo/fyi/blob/master/expo-extension-migration.md):

```js
let isExpo = false;
try {
  const Constants = require("expo-constants");
  // True if the app is running in an `expo build` app or if it's running in Expo Go.
  isExpo =
    Constants.executionEnvironment === "standalone" ||
    Constants.executionEnvironment === "storeClient";
} catch {}
```

This code explodes Webpack in another new way though. A few actually - all related to expecting packages that don't exist in this context.

```
[0] ERROR in ./node_modules/@unimodules/react-native-adapter/build/nativeEmitters.js 1:0-93
[0] Module not found: Error: Can't resolve 'react-native-web/dist/vendor/react-native/emitter/EventEmitter' in 'C:\projects\project\node_modules\@unimodules\react-native-adapter\build'
[0] resolve 'react-native-web/dist/vendor/react-native/emitter/EventEmitter' in 'C:\projects\project\node_modules\@unimodules\react-native-adapter\build'
[0]   Parsed request is a module
[0]   using description file: C:\projects\project\node_modules\@unimodules\react-native-adapter\package.json (relative path: ./build)
[0]     resolve as module
[0]       looking for modules in C:\projects\project
[0]         C:\projects\project\react-native-web doesn't exist
[0]       C:\projects\project\node_modules\@unimodules\react-native-adapter\build\node_modules doesn't exist or is not a directory
[0]       C:\projects\project\node_modules\@unimodules\react-native-adapter\node_modules doesn't exist or is not a directory
[0]       C:\projects\project\node_modules\@unimodules\node_modules doesn't exist or is not a directory
[0]       C:\projects\project\node_modules\node_modules doesn't exist or is not a directory
[0]       looking for modules in C:\projects\project\node_modules
[0]         C:\projects\project\node_modules\react-native-web doesn't exist
[0]       C:\projects\node_modules doesn't exist or is not a directory
[0]       C:\node_modules doesn't exist or is not a directory
[0]  @ ./node_modules/@unimodules/react-native-adapter/build/SyntheticPlatformEmitter.js 2:0-51 9:28-43
[0]  @ ./node_modules/@unimodules/react-native-adapter/build/index.js 6:0-66 9:0-11:35
[0]  @ ./node_modules/@unimodules/core/build/AdapterProxy.js 1:0-49 1:0-49
[0]  @ ./node_modules/@unimodules/core/build/index.js 1:0-31 1:0-31
[0]  @ ./node_modules/expo-constants/build/Constants.js 1:0-66 11:4-34 13:8-47 14:26-65 16:13-58 17:37-82 86:22-32
[0]  @ ./packages/shared/project-utils/src/environment.js 7:20-45
[0]  @ ./app/desktop/preload/sharedPreload.js 10:0-56 12:17-31 13:15-29 15:4-18 20:14-29 58:2-5
[0]
[0] ERROR in ./node_modules/react-native/index.js 13:7
[0] Module parse failed: Unexpected token (13:7)
[0] You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
[0] | 'use strict';
[0] |
[0] > import typeof AccessibilityInfo from './Libraries/Components/AccessibilityInfo/AccessibilityInfo';
[0] | import typeof ActivityIndicator from './Libraries/Components/ActivityIndicator/ActivityIndicator';
[0] | import typeof Button from './Libraries/Components/Button';
[0]  @ ./node_modules/expo-constants/build/Constants.js 2:0-55 24:4-31 26:8-50 27:41-83 77:12-23
[0]  @ ./packages/shared/project-utils/src/environment.js 7:20-45
[0]  @ ./app/desktop/preload/sharedPreload.js 10:0-56 12:17-31 13:15-29 15:4-18 20:14-29 58:2-5
[0]

```

So the only option I see left is one I really don't want to do - define `isElectron` using `webpack.DefinePlugin` - because that's very rigid and we already have a proper way to detect Electron, so really what we'd be defining is `isNotExpo`, which seems all sorts of convoluted.

Maybe I'm the only one, but I miss the globals. I get the desire to remove them, but I think they were removed too quickly. It feels like an hour that didn't need to be burned, but maybe I've overlooked something simple? Any advice?

PS: We're using a monorepo. Thanks for reading, this comment just kept getting longer.

**Comment by Slapbox** [comment]
So it looks like we're going to have to go the route of using `webpack.DefinePlugin` to define `global.isNotExpo`.

Unless I'm overlooking something, at least in my testing, the suggested method for handling this of using `require()` with Webpack5 means that if we need to `export` the value, it's going to export `false` even if the `require`'d file would later redefine that to `true`.

We even tried using [babel-plugin-transform-define](https://github.com/FormidableLabs/babel-plugin-transform-define/) in our `babelrc.js` file in our mobile app's folder, but we can't get that working either.

So it seems like we're stuck either defining `global.isNotExpo` or rewriting hundreds of `import`s to `require`s, and I'm not in love with either of these ideas. For now I guess it's defining `global.isNotExpo` so we can move on, but I'm wondering if maybe some better solution could be conceived of?

**Pull request merged**